src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java

```diff
@@ -0,0 +1,115 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.randomized;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import se.liu.ida.hefquin.engine.queryplan.LogicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizationException;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizer;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.CostModel;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.QueryOptimizationContext;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RewritingRule;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RuleApplication;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RuleInstances;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.utils.CostEstimationUtils;
+import se.liu.ida.hefquin.engine.utils.Pair;
+
+public class IterativeImprovementBasedQueryOptimizer implements QueryOptimizer {
+	
+	protected final QueryOptimizationContext context;
+	protected final StoppingConditionForIterativeImprovement condition;
+	protected final Random rng;
+
+	
+	public IterativeImprovementBasedQueryOptimizer (final QueryOptimizationContext ctxt, final StoppingConditionForIterativeImprovement x) {
+		assert ctxt != null;
+		assert x != null;
+		context = ctxt;
+		condition = x;
+		rng = new Random();
+	}
+	
+	
+	@Override
+	public PhysicalPlan optimize( final LogicalPlan initialPlan ) throws QueryOptimizationException {
+		return optimize( context.getLogicalToPhysicalPlanConverter().convert(initialPlan,false) );
+	}
+
+	public PhysicalPlan optimize( final PhysicalPlan initialPlan ) throws QueryOptimizationException {
+		// The best plan we have found so far. As we have only found one plan, it is the best one so far.
+		PhysicalPlan bestPlan = initialPlan;
+		
+		// Creating the completable future.
+		CostModel costModel = context.getCostModel(); //Created as a value and stored rather than just using the getCostModel, because the same model is used later too.
+
+		Double[] initialCost = CostEstimationUtils.getEstimates(costModel, initialPlan);
+		
+		// generation = number of times the outer loop has run. Has to be declared here since it will increment each outer loop.
+		int generation = 0;
+		
+		// bestCost = best possible cost out of everything that the algorithm has found. Has to be declared here so that we have something to compare to in all the iterations of the outer loop without losing track of it between loops.
+		Double bestCost = initialCost[0];

```

Comment:
As I posted before, the IDE gives me an error if I do it like this. It tells me to either make it a Double[] or to change the getEstimates function to return a Double instead of a Double[].

Reply:
Did you notice that `[0]` that I have added at the end of the line? With this, it should directly access the first element of the array returned by `getEstimates` and assign that element to the `Double`. This doesn't work in your IDE?

Reply:
Oh, I didn't notice the [0], sorry! I didn't even know you could do something like that. That's neat, thanks!